### DISCUSSION TITLE
Alpha-sort language files when touched?

### DISCUSSION THREAD

**Discussion by torvista** [zencart/zencart#6677] [Ideas]
Mostly are currently in a logical order, grouped as the texts may appear on the page.
However if making significant additions/removals/reordering (but not sorted) in a file eg. the last one for layout_settings (https://github.com/zencart/zencart/pull/6630/files#diff-96a4f4eca425efc9b6dabb381c591e96ef7fd8cfaf262745a487fa536a2208ee), it's extra work to rejig an existing translation, to move things around one by one so they match up/not lose existing texts, when if it was alpha-sorted it would be easy.

Also, making additions to an alpha-sorted file is easier, no decisions to make about location!

**Comment by drbyte** [comment]
Point noted. Outcome undecided.

Deciding where to position translation strings is a challenge: Yes, alpha-sorting by key makes "some" things easier. But it completely loses context. Positioning related strings together makes it super fast to translate related concepts without wondering "how did I translate it up above?" when translating a new file or a large section of additions/changes.

**Reply by torvista** [reply]
In an ideal world, similar translations would have similarly-named constants and so would be adjacent. Meanwhile, in the real world.....

**Reply by drbyte** [reply]
Fair point. 
The convention used has often been that the first word of the constant name is more about its usage than its topic. eg: TEXT_HEADING_FOO_* relates to text used for headings and `FOO` changes for whatever topic the heading is related to.

Unfortunately, like you say, that doesn't lead to the easy sort-by-alpha to keep contextual strings together.

**Reply by torvista** [reply]
...and it leads to inconsistent-looking constant usage in the code, just when an existing constant fits the text required.

Perhaps everything should be TEXT_

That would make naming of new strings simpler/quicker if context was not considered.

**Reply by drbyte** [reply]
> Perhaps everything should be TEXT_
That would make naming of new strings simpler/quicker if context was not considered

I hear you. And yet there are limitations to that idea.

`TEXT_SPECIALS => 'Specials'` which, without desiring any specificity, could be used for both page-header and a table column header, but also as a form field label, or as a sidebox header
vs
`TEXT_PAGE_HEADER_SPECIALS => 'Store Specials'`
`TEXT_TABLE_HEADING_SPECIALS => 'For Specials?` (eg: a switch for whether a thing applies when a Special is active)
`TEXT_SIDEBOX_HEADER_SPECIALS => 'Current Specials'`

**Reply by scottcwilson** [reply]
... and making large changes to refactor the naming conventions would be a NIGHTMARE for storeowners, because now an undefined constant leads to a white page. 

I sorted the definitions for some language files when we moved to the new language file format so I could look for redundancies and consolidation opportunities, not because it was inherently better to have them sorted.  As Chris noted, the downside of doing this is the loss of context for adjacently used defintions.   We should probably leave them alone at this point so we can go back to having smaller diffs between releases.
